packages/localize/src/program-analysis.ts

```diff
@@ -86,112 +88,250 @@ function extractMsg(
+    return {
+      error: createDiagnostic(
+        file,
+        node,
+        `Expected 1 or 2 arguments to msg(), got ${node.arguments.length}`
+      ),
+    };
   }
-  const [nameArg, contentsArg] = node.arguments;
-  if (!isStaticString(nameArg) || !nameArg.text) {
-    return createDiagnostic(
-      file,
-      nameArg,
-      `Expected first argument to msg() to be a static and non-empty string`
-    );
+  const [templateArg, optionsArg] = node.arguments;
+
+  const templateResult = extractTemplate(templateArg, file);
+  if (templateResult.error) {
+    return {error: templateResult.error};
   }
-  const name = nameArg.text;
+  const {contents, params, isLitTemplate} = templateResult.result;
 
-  if (isStaticString(contentsArg)) {
-    // E.g. msg('foo', 'bar')
-    return {
+  const optionsResult = extractOptions(optionsArg, file);
+  if (optionsResult.error) {
+    return {error: optionsResult.error};
+  }
+  const options = optionsResult.result;
+  if (options.id === undefined) {
+    // TODO(aomarks) Implement.
+    throw new Error('Not yet implemented: auto ID');
+  }
+  const name = options.id;
+
+  return {
+    result: {
       name,
-      contents: [contentsArg.text],
-      node,
       file,
+      node,
+      contents,
+      params,
+      isLitTemplate,
       descStack: descStack.map((desc) => desc.text),
-      isLitTemplate: false,
+    },
+  };
+}
+
+/**
+ * Analyze the options argument to a msg call.
+ */
+export function extractOptions(
+  node: ts.Node | undefined,
+  file: ts.SourceFile
+): ResultOrError<
+  {id?: string; args?: ts.NodeArray<ts.Expression>},
+  ts.Diagnostic
+> {
+  if (node === undefined) {
+    return {result: {}};
+  }
+  if (!ts.isObjectLiteralExpression(node)) {
+    return {
+      error: createDiagnostic(
+        file,
+        node,
+        `Expected second argument to msg() to be an object`
+      ),
     };
   }
 
-  if (isLitTemplate(contentsArg)) {
-    if (ts.isNoSubstitutionTemplateLiteral(contentsArg.template)) {
-      // E.g. msg('foo', html`bar <b>baz</b>`)
+  let id: string | undefined = undefined;
+  let args: ts.NodeArray<ts.Expression> | undefined = undefined;
+
+  for (const property of node.properties) {
+    if (!ts.isPropertyAssignment(property)) {
       return {
-        name,
-        contents: combineAdjacentPlaceholders(
-          replaceHtmlWithPlaceholders(contentsArg.template.text)
+        error: createDiagnostic(
+          file,
+          property,
+          `Options shorthand object assignment is not supported`
+        ),
+      };
+    }
+
+    let name;
+    if (ts.isIdentifier(property.name)) {
+      name = property.name.escapedText;
+    } else if (ts.isStringLiteral(property.name)) {
+      name = property.name.text;
+    } else {
+      return {
+        error: createDiagnostic(
+          file,
+          property.name,
+          `Options shorthand object assignment must be identifier or string literal`
```

Comment:
Is this message correct? Here would would have a computed property name I think.

Reply:
Oops yeah, fixed. I updated both errors to be more clear too.